File: python/semanticabi/abi/item/Explode.py
from __future__ import annotations
from dataclasses import dataclass
from functools import cached_property
from typing import List, TypedDict, Optional, Dict

from semanticabi.abi.InvalidAbiException import InvalidAbiException
from semanticabi.abi.item.SemanticParameter import SemanticParameters, SemanticParameter


class TypedExplode(TypedDict):
    paths: List[str]


@dataclass
class Explode:
    """
    Defines the paths in a jsonpath-like dot-separated format to explode into separate rows.
    """

    @staticmethod
    def from_json(json: TypedExplode) -> Explode:
        return Explode(json['paths'])

    paths: List[str]

    @cached_property
    def path_parts(self) -> Dict[str, List[str]]:
        """
        The path parts for each path
        """
        return {path: path.split('.') for path in self.paths}

    def validate(self, all_parameters: List[SemanticParameters]):
        for path, path_parts in self.path_parts.items():
            parameter: Optional[SemanticParameter] = None

            # Find the root parameter in the input or output traces
            for parameters_set in all_parameters:
                parameter = parameters_set.parameter(path_parts[0]) if parameters_set.has_parameter(path_parts[0]) else None
                if parameter is not None:
                    break

            for part in path_parts[1:]:
                if parameter is None:
                    break

                if parameter.exclude:
                    raise InvalidAbiException(f'Explode path \'{path}\' cannot reference an excluded parameter: {part}')

                if parameter.parameter.is_array:
                    raise InvalidAbiException(f'Explode path \'{path}\' does not support nested arrays: {part}')

                parameter = parameter.components.parameter(part) if parameter.components is not None and parameter.components.has_parameter(part) else None

            if parameter is None:
                raise InvalidAbiException(f'Explode path \'{path}\' not found in item parameters')

            if not parameter.parameter.is_array:
                raise InvalidAbiException(f'Explode path \'{path}\' is not an array')

            if parameter.parameter.is_array_of_arrays:
                raise InvalidAbiException(f'Explode path \'{path}\' does not support array of arrays')

File: js/src/semanticabi/abi/item/semantic/Explode.ts
import {SemanticParameter, SemanticParameters} from "semanticabi/abi/item/semantic/SemanticParameters";
import {SemanticItemTransform} from "semanticabi/abi/item/semantic/SemanticItemTransform";
import {SemanticAbiValidationError} from "semanticabi/abi/item/semantic/SemanticAbiValidateable";
import {Left, Right} from "common/Either";
import {Optional} from "common/Optional";
import {AbiItem} from "semanticabi/abi/item/AbiItem";

export type ExplodeJson = {
    paths: string[]
}

/**
 * Explode the parameters into their parts.
 */
export class Explode implements SemanticItemTransform {

    public readonly pathParts: Map<string, string[]>;

    constructor(
        public readonly paths: string[]
    ) {
        this.pathParts = new Map(paths.map(path => [path, path.split('.')] as [string, string[]]));
    }

    validate(item: AbiItem, inputAndOutputs: SemanticParameters[]): Optional<SemanticAbiValidationError> {
        for (const [path, pathParts] of this.pathParts) {
            const pathValidation = inputAndOutputs.find(parameters => parameters.has(pathParts[0]))
                .getAt(pathParts)
                // a couple more checks on the resolved parameter
                .flatMap<SemanticParameter>(parameter => {
                    if (parameter.exclude) {
                        return new Left(new SemanticAbiValidationError(`Explode path '${path}' cannot reference an excluded parameter.`));
                    }
                    if (!parameter.parameter.isArray || parameter.parameter.isArrayOfArrays) {
                        return new Left(new SemanticAbiValidationError(`Explode path '${path}' is not a one dimensional array that can be exploded.`));
                    }

                    return new Right(parameter);
                })
                // swap to return right as the optional error
                .swap()
                .optional();

            if (pathValidation.isPresent) {
                return pathValidation;
            }
        }

        return Optional.none();
    }

    static fromJSON(json: ExplodeJson): Explode {
        return new Explode(json['paths']);
    }

}